### ISSUE TITLE
Style/HashConversion auto-correct missed parantheses for OR statements

### ISSUE THREAD

**Issue by jswk** [rubocop/rubocop#10305]
The `Style/HashConversion` cop changes OR semantics under some conditions.

--------

## Expected behavior

Given the following code:
```ruby
Hash[a || b]
```

RuboCop is expected to auto-correct it to:
```ruby
(a || b).to_h
```

## Actual behavior

RuboCop auto-corrects it to:
```ruby
a || b.to_h
```

Now `a` will be returned without conversion to Hash, given it's not nil.

## Steps to reproduce the problem

Paste the first snipped in a file, and run `rubocop --auto-correct` on it.

## RuboCop version

```
$ rubocop -V
1.23.0 (using Parser 3.0.3.2, rubocop-ast 1.15.0, running on ruby 3.0.2 x86_64-linux)
```
